src/coreclr/jit/fgdiagnostic.cpp

```diff
@@ -3278,7 +3278,7 @@ void Compiler::fgDebugCheckTypes(GenTree* tree)
             //    * All kinds of locals
             //    * GT_COMMA wrapped around any of the above.
             //
-            if (node->TypeIs(TYP_ULONG, TYP_UINT))
+            if (node->TypeIs(TYP_ULONG, TYP_UINT) && !node->OperIs(GT_CAST))
```

Comment:
We shouldn't be changing this, it will require a lot more complex changes in the IR to support.

Whether its doing a signed conversion is instead supported via the additional `unsigned` metadata tracked by the cast nodes

Reply:
Can you explain your comment here? I am seeing some tests crash here when checking for node type during conversion. Are you saying a separate check is needed or we cannot add this check at all?

```
Assert failure(PID 75840 [0x00012840], Thread: 26876 [0x68fc]): Assertion failed '!"TYP_ULONG and TYP_UINT are not legal in IR"' in 'VectorTest+VectorConvertTest:VectorConvertSingleUInt(System.Numerics.Vector`1[float]):int' during 'Post-Morph' (IL size 177; hash 0xd1d1e19a; Instrumented Tier0)

    File: D:\Git_repos\runtime\src\coreclr\jit\fgdiagnostic.cpp:3284
    Image: D:\Git_repos\runtime\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe


Return code:      1
Raw output file:      D:\Git_repos\runtime\artifacts\tests\coreclr\windows.x64.Debug\JIT\SIMD\VectorConvert_ro_Target_64Bit\output.txt
Raw output:
BEGIN EXECUTION

D:\Git_repos\runtime\artifacts\tests\coreclr\windows.x64.Debug\JIT\SIMD\VectorConvert_ro_Target_64Bit>set DOTNETCLI=D:\Git_repos\runtime\dotnet.cmd
 "D:\Git_repos\runtime\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe" -p "System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true"  VectorConvert_ro_Target_64Bit.dll
               [000202] -A-XG+-----                         *  CAST      uint <- float
               [000201] -A-XG+-----                         \--*  HWINTRINSIC float  float ToScalar
               [000200] -A-XG+-----                            \--*  HWINTRINSIC simd16 float FixupScalar
               [000197] -A-XG+-----                               +--*  COMMA     float
               [000195] DA-XG+-----                               |  +--*  STORE_LCL_VAR float  V17 tmp10
               [000026] ---XG+-----                               |  |  \--*  HWINTRINSIC float  float GetElement
               [000203] n---G+-----                               |  |     +--*  IND       simd32
               [000021] -----+-----                               |  |     |  \--*  LCL_VAR   byref  V00 arg0
               [000025] ---X-+-----                               |  |     \--*  COMMA     int
               [000024] ---X-+-----                               |  |        +--*  BOUNDS_CHECK_ArgRng void
               [000023] -----+-----                               |  |        |  +--*  LCL_VAR   int    V04 loc3
               [000022] -----+-----                               |  |        |  \--*  CNS_INT   int    8
               [000020] -----+-----                               |  |        \--*  LCL_VAR   int    V04 loc3
               [000196] -----+-----                               |  \--*  LCL_VAR   float  V17 tmp10
               [000198] -----+-----                               +--*  LCL_VAR   float  V17 tmp10
               [000194] -----+-N---                               +--*  CNS_VEC   simd16<0x08000088, 0x00000000, 0x00000000, 0x00000000>
               [000199] -----+-N---                               \--*  CNS_INT   int    0
```

Reply:
We shouldn't be adding this check at all. It's intentional that nodes don't return `TYP_UINT` or `TYP_ULONG`.

Instead, they always return `TYP_INT` or `TYP_LONG` and where sign is important (such as for conversions) there exists a separate field on the node that identifies the type it comes from vs the actual target type.

The create cast node API is `GenTreeCast* gtNewCastNode(var_types typ, GenTree* op1, bool fromUnsigned, var_types castType)` and so for `float->uint`, we'd want `gtNewCastNode(TYP_INT, op1, false, TYP_UINT)`, which produces a node that looks like:
```
[000001] -----------                         \--*  CAST      int <- uint <- float
```

This is in constrast to `float->int` which is `gtNewCastNode(TYP_INT, op1, false, TYP_INT)` and produces a node that looks like:
```
[000001] -----------                         \--*  CAST      int <- float
```

Notably the handling in `importer.cpp` should already be doing this, so it just needs to be repeated in places like morph where you're inserting the alternative sequences.

Reply:
Hi Tanner,

Thanks for the explanation. I re-read your first comment a couple of times and was able to figure it out. I have made this changes and will push a clean change with changes for all the other comments today.